substrate/validator-sets/pallet/src/lib.rs

```diff
@@ -0,0 +1,206 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[frame_support::pallet]
+pub mod pallet {
+  use scale::{Encode, Decode};
+  use scale_info::TypeInfo;
+
+  use frame_system::pallet_prelude::*;
+  use frame_support::pallet_prelude::*;
+
+  use validator_sets_primitives::*;
+
+  #[pallet::config]
+  pub trait Config: frame_system::Config + TypeInfo {
+    type RuntimeEvent: IsType<<Self as frame_system::Config>::RuntimeEvent> + From<Event<Self>>;
+  }
+
+  #[pallet::genesis_config]
+  #[derive(Clone, PartialEq, Eq, Debug, Encode, Decode, MaxEncodedLen)]
+  pub struct GenesisConfig<T: Config> {
+    /// Bond requirement to join the initial validator set.
+    /// Every participant at genesis will automatically be assumed to have this much bond.
+    /// This bond cannot be withdrawn however as there's no stake behind it.
+    pub bond: Amount,
+    /// Amount of coins to spawn the network with in the initial validator set.
+    pub coins: Coin,
+    /// List of participants to place in the genesis set.
+    pub participants: Vec<T::AccountId>,
+  }
+
+  #[cfg(feature = "std")]
+  impl<T: Config> Default for GenesisConfig<T> {
+    fn default() -> Self {
+      GenesisConfig { bond: Amount(0), coins: Coin(0), participants: vec![] }
+    }
+  }
+
+  // Max of 16 coins per validator set
+  // At launch, we'll have BTC, ETH, DAI, and XMR
```

Comment:
We have to include tokens here. It's a list of the coins managed and each token has a distinct token impl on Serai, classifying it as a coin. There's no way around it.

As for this "max", it is a hard limit, yet the reason we need a hard limit is for gas estimation purposes. If we set it to 32, then it'll be estimated as 32 elements long when it's likely 4. The max I'd see is 8 and I went with 16 just to be cautious.

Finally, no, we can't guard tokens by contracts AFAIK. Mind elaborating there? Even if we did a light client, it wouldn't void bond requirements.

Reply:
When set members outgrow dPoS counts we will be in a situation where we can form super-set consensus to manage token contract ownership versus a single set. I think this security can result in reduced or eliminated assets for tokens that can be guarded. As you mentioned, the other sets would need observation on the relevant chains to be a party to this contract but in theory, this could work with light client driven watchers in other sets as you already have one BFT set forming consensus around full nodes.

Reply:
So currently, the idea is the chain itself is validated by a super-set. In the future discussions, the chain may be validated by a dedicated set.

In order for tokens to not be under a multisig, we'd need a light client. Light clients would effectively be a multisig of whoever produces consensus commits for the Serai blockchain. That maintains a bond requirement on either the super-set or the dedicated set. Else, the dedicated set can forge consensus messages and...

Right now, they can drain coins externally and SRI in pools, yet we dismiss that SRI having further value (since the coins in the pools were drained) AND their bond outweighs the externally stolen coins. Under a light client scheme, we lose bond despite having the same risk architecture.